### PR TITLE
RavenDB-15645 avoid segment split for different number of values

### PR DESCRIPTION
**Pull request by karmeli87**
If we append 10 entries when the first has 3 values and the rest only one,
we would call segment split 9 times.

### CODE REVIEW

File: src/Raven.Server/Documents/TimeSeries/SegmentHeader.cs [hunk 1]
```diff
@@ -18,6 +18,14 @@ public unsafe struct SegmentHeader
...
-        public fixed byte Reserved[2];
+        public SegmentVersion Version;
+        [FieldOffset(15)]
+        public fixed byte Reserved[1];
+    }
+
+    public enum SegmentVersion : byte
+    {
+        V50000 = 0,
+        V50001 = 1
```
Comment (ppekrol) [inline src/Raven.Server/Documents/TimeSeries/SegmentHeader.cs]: can we also have `Current = V50001` and use that in the Initialize?

Comment (karmeli87) [reply src/Raven.Server/Documents/TimeSeries/SegmentHeader.cs]: done

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs [hunk 2]
```diff
@@ -2038,6 +2037,26 @@ private static void EnsureStatsAndDataIntegrity(DocumentsOperationContext contex
...
+                foreach (var state in segment.SegmentValues.Span)
+                {
+                    Debug.Assert(double.IsNaN(state.First) == false, "First is NaN");
+                    Debug.Assert(double.IsNaN(state.Last) == false, "Last is NaN");
+                    Debug.Assert(double.IsNaN(state.Min) == false, "Min is NaN");
+                    Debug.Assert(double.IsNaN(state.Max) == false, "Max is NaN");
+                    Debug.Assert(double.IsNaN(state.Sum) == false, "Sum is NaN");
+                    Debug.Assert(double.IsNaN(state.Count) == false, "Count is NaN");
+                    Debug.Assert(state.Count > 0, "Count is Zero");
+                }
```
Comment (ayende) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs:2052]: Why do we have all of those asserts here? If we are validating this, let's move all of this to a `[Conditional("DEBUG")]` method.

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs [hunk 3]
```diff
@@ -2038,6 +2037,26 @@ private static void EnsureStatsAndDataIntegrity(DocumentsOperationContext contex
...
+
+                if (name.Contains(TimeSeriesConfiguration.TimeSeriesRollupSeparator))
+                {
+                    var noNaN = segment.YieldAllValues(context, baseline: default, includeDead: false).All(x => x.Values.ToArray().All(y => double.IsNaN(y) == false));
+                    Debug.Assert(noNaN, "Rollup has NaN");
+                }
```
Comment (ayende) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs:2058]: Same here, this is debug only code, no?

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs [hunk 4]
```diff
@@ -468,7 +473,13 @@ public IEnumerable<SingleResult> YieldAllValues(JsonOperationContext context, By
                     if (status == Dead && includeDead == false)
                         continue;
 
-                    var cur = baseline.AddMilliseconds(ts);
+                    var next = baseline.AddMilliseconds(ts);
+                    if (next == current && 
+                        Version == SegmentVersion.V50000) // fix legacy issue RavenDB-15617
```
Comment (ayende) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs:478]: What is the difference between the versions?

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs [hunk 5]
```diff
@@ -569,6 +580,13 @@ public Span<byte> AsSpan()
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsValid(ulong status, long value)
+        {
+            // valid = live and not NaN
+            return status == Live && (value & 0x7FFFFFFFFFFFFFFF) > 0x7FF0000000000000 == false;
```
Comment (ayende) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs:587]: wouldn't it be better to use `double.IsNaN(Convert.LongToDoubleBits(value))` ?

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs [hunk 6]
```diff
@@ -747,6 +772,23 @@ private static void ThrowValuesOutOfRange(int numberOfValues)
...
+                return false;
+
+            if (NumberOfLiveEntries == 0)
+                return false;
+
+            foreach (var value in SegmentValues.Span)
+            {
+                if (double.IsNaN(value.Last))
+                    return true;
+            }
```
Comment (ayende) [inline src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs:791]: Why do we check all the items in the array?

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs (hunk 2 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs:2052]: It is all under `[Conditional("DEBUG")]`

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs (hunk 5 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs:587]: it is the impl of `double.IsNaN`, so I avoid the double conversion (from `long` to `double` and then back from `double` to `long`)

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs (hunk 4 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs:478]: different issue...  we might have duplicate timestamps :/

File: src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs (hunk 6 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs:791]: Eventually this check is used to decide whether we can use the `StatefulTimestampValueSpan` of the segment or not.

Comment (ayende) [reply src/Raven.Server/Documents/TimeSeries/TimeSeriesValuesSegment.cs:791]: can you add a comment explaining the reasoning

**Review by ayende — APPROVED** [review]

**Comment by ayende** [comment]
Need schema upgrade

**Comment by ayende** [comment]
We are reusing a reserved value. We don't _have_ to do anything, but it is a concern, no?

**Comment by karmeli87** [comment]
Yes, we basically need only to bump the version

**Pull request merged**